crates/ty_server/src/server/main_loop.rs

```diff
@@ -25,19 +26,48 @@ impl Server {
 
             match next_event {
                 Event::Message(msg) => {
-                    if self.connection.handle_shutdown(&msg)? {
-                        break;
-                    }
+                    let client = Client::new(
+                        self.main_loop_sender.clone(),
+                        self.connection.sender.clone(),
+                    );
+
                     let task = match msg {
                         Message::Request(req) => {
                             self.session
                                 .request_queue_mut()
                                 .incoming_mut()
                                 .register(req.id.clone(), req.method.clone());
 
+                            if self.session.is_shutdown_requested() {
+                                tracing::warn!(
+                                    "Received request after server shutdown was requested, discarding"
+                                );
+                                client.respond_err(
+                                    req.id,
+                                    lsp_server::ResponseError {
+                                        code: lsp_server::ErrorCode::InvalidRequest as i32,
+                                        message: "Shutdown already requested.".to_owned(),
+                                        data: None,
+                                    },
+                                )?;
+                                continue;
+                            }
+
                             api::request(req)
                         }
-                        Message::Notification(notification) => api::notification(notification),
+                        Message::Notification(notification) => {
+                            if notification.method == lsp_types::notification::Exit::METHOD {
+                                tracing::debug!("Received exit notification, exiting");
+                                if !self.session.is_shutdown_requested() {
+                                    tracing::warn!(
+                                        "Received exit notification before shutdown request"
+                                    );
+                                }
```

Comment:
I'd consider these two messages to be mutually exclusive, so maybe something like the following (appended "exiting" to both as the server would exit in both scenario):
```suggestion
                                if self.session.is_shutdown_requested() {
                                	tracing::debug!("Received exit notification, exiting");
                                } else {
                                    tracing::warn!(
                                        "Received exit notification before shutdown request, exiting"
                                    );
                                }
```

Reply:
I don't think they're exlusive to each other. We always exit, but we want a warning when we do so before we receive a shutdown request. We could probably return with an `Err` in that case, similar to how we error when we didn't receive an `ExitNotification`

Reply:
Returning with an error is more conform with the LSP specification

> A notification to ask the server to exit its process. The server should exit with success code 0 if the shutdown request has been received before; otherwise with error code 1.